compiler/src/iree/compiler/Codegen/Dialect/GPU/Transforms/Transforms.cpp

```diff
@@ -288,6 +288,322 @@ LogicalResult fuseForallIntoConsumer(RewriterBase &rewriter,
+                                        offsets.begin() + group.back() + 1);
+    SmallVector<Value> offsetVals =
+        llvm::map_to_vector(groupOffsets, [&](OpFoldResult ofr) {
+          return getValueOrCreateConstantIndexOp(rewriter, loc, ofr);
+        });
+    OpFoldResult collapsedOffset =
+        rewriter.create<affine::AffineLinearizeIndexOp>(loc, offsetVals, basis)
+            .getResult();
+    ArrayRef<OpFoldResult> groupSizes(sizes.begin() + group.front(),
+                                      sizes.begin() + group.back() + 1);
+    OpFoldResult collapsedSize = prod(groupSizes);
+    newOffsets.push_back(collapsedOffset);
+    newSizes.push_back(collapsedSize);
+  }
+  SmallVector<OpFoldResult> newStrides(newSizes.size(),
+                                       rewriter.getIndexAttr(1));
+
+  // Collapse the slice source.
+  loc = parallelInsertOp.getParallelCombiningParent()->getLoc();
+  rewriter.setInsertionPoint(parallelInsertOp.getParallelCombiningParent());
+  auto newCollapse = rewriter.create<tensor::CollapseShapeOp>(
+      loc, parallelInsertOp.getSource(), reassociations);
+
+  // Collapse the parallel insert slice.
+  rewriter.setInsertionPoint(parallelInsertOp);
+  auto newInsertOp = rewriter.replaceOpWithNewOp<tensor::ParallelInsertSliceOp>(
+      parallelInsertOp, newCollapse, parallelInsertOp.getDest(), newOffsets,
+      newSizes, newStrides);
+  return newInsertOp;
+}
+
+FailureOr<scf::ForallOp>
+fuseCollapseShapeIntoProducerForall(RewriterBase &rewriter,
+                                    scf::ForallOp forallOp,
+                                    tensor::CollapseShapeOp collapseOp) {
+  // Check that there is a single user of the collapsed result.
+  auto forallResult = cast<OpResult>(collapseOp.getSrc());
+  if (!forallResult.hasOneUse()) {
+    return rewriter.notifyMatchFailure(forallOp,
+                                       "forall result has multiple uses");
+  }
+
+  // Get the result's corresponding parallel_insert_slice op.
+  SmallVector<Operation *> parallelInsertOps = forallOp.getCombiningOps(
+      forallOp.getRegionIterArgs()[forallResult.getResultNumber()]);
+  if (parallelInsertOps.size() != 1) {
+    return rewriter.notifyMatchFailure(
+        forallOp, "Expected a single parallel_insert_slice");
+  }
+
+  auto parallelInsertOp =
+      dyn_cast<tensor::ParallelInsertSliceOp>(parallelInsertOps.front());
+  if (!parallelInsertOp) {
+    return rewriter.notifyMatchFailure(
+        forallOp, "Expected parallel_insert_slice combining op");
+  }
+
+  // Collect all collapsable users of the loop init argument, and verify that
+  // they are all slices with the same offsets, sizes, and (unit) strides.
+  SmallVector<ReassociationIndices> reassociations =
+      collapseOp.getReassociationIndices();
+  Value loopInit = parallelInsertOp.getDest();
+  FailureOr<SmallVector<Operation *>> maybeCollapsableOps =
+      collectCollapsableUsers(rewriter, loopInit, reassociations);
```

Comment:
Rather than creating a SmallVector here, have this return a LogicalResult and just use `loopInit.getUsers()` in subsequent uses.